controllers/cluster_controller.go

```diff
@@ -287,3 +295,45 @@ func splitMachineList(list *clusterv1.MachineList) (*clusterv1.MachineList, *clu
 	}
 	return controlplanes, nodes
 }
+
+// reconcileClusterStatus reconciles the status on the Cluster associated with Machines.
+func (r *ClusterReconciler) reconcileClusterStatus(o handler.MapObject) []ctrl.Request {
+	result := []ctrl.Request{}
+
+	m, ok := o.Object.(*clusterv1.Machine)
+	if !ok {
+		r.Log.Error(errors.New("did not get machine"), "got", fmt.Sprintf("%T", o.Object))
+		return result
+	}
+
+	cluster, err := util.GetClusterFromMetadata(context.Background(), r.Client, m.ObjectMeta)
+	if err != nil {
+		r.Log.Error(err, "failed to get cluster", "machine", m.Name, "cluster", m.ClusterName)
+		return result
+	}
+
+	// set the Status.ControlPlaneInitialized on the Cluster.
+	if util.IsControlPlaneMachine(m) && m.Status.NodeRef != nil {
+		if !cluster.Status.ControlPlaneInitialized {
+			patchHelper, err := patch.NewHelper(cluster, r)
+			if err != nil {
+				r.Log.Error(err, "failed to create patch helper", "cluster", cluster.Name, "namespace", cluster.Namespace)
+				return result
+			}
+			cluster.Status.ControlPlaneInitialized = true
+			if err := patchHelper.Patch(context.Background(), cluster); err != nil {
```

Comment:
A MapFunc should not perform any logic other than mapping the watched resource (Machines, in this case) to an instance of the resource that the reconciler handles (Cluster, in this case). Would you please split this logic into 2 functions: 1 for the MapFunc, and 1 for using `r.Client.Get()` to get a list of machines for the cluster and set ControlPlaneInitialized accordingly?

Reply:
thanks for your review! I've updated the pr, can you take a look?